checker/jdk/nullness/src/java/util/AbstractList.java

```diff
@@ -181,7 +181,7 @@ public E remove(int index) {
      * @throws NullPointerException {@inheritDoc}
      */
     @Pure
-    public int indexOf(@Nullable Object o) {
+    public int indexOf(Object o) {
```

Comment:
At a high level, I see 3 cases:

1. the interfaces themselves (like `List`)
2. the skeletal abstract classes (like `AbstractList`)
3. the concrete implementations (like `ArrayList`)

For (1), we're here exactly because we're aiming to make conservative assumptions by default (though allow them to be loosened with stub files). So we don't want `@Nullable` on parameters there.

Jumping ahead to (3), I would have guessed that classes like `ArrayList` _would_ have `@Nullable` on their parameters, since their implementations support null inputs and their docs reflect this (and the Checker Framework supports generalizing parameter types). I can see a couple possible arguments against that, though I have reservations about each:

- First, users rarely store objects in variables/fields whose static type is `ArrayList`, anyway, so the annotations on `ArrayList` itself rarely matter. (The argument about static types might become somewhat less true as users adopt `var`.)
- Second, users can extend `ArrayList` and override this method to reject null inputs. Taken to the extreme, this seems like an argument against putting `@Nullable` on a parameter of _any_ extensible method -- and an argument _for_ putting `@Nullable` on the _return value_ of every extensible method (since anyone might override it to return `null`). So I think at some point we have to take a stand on which of those overrides are valid and which aren't. Given that `ArrayList` doesn't document that its methods throw `NullPointerException`, I'd expect us to permit `null` here (unless the other argument above this one holds sway).

Coming back to (2), I think that's more like (1): Even though a given implementation might support null, the classes that override it might not, and the docs reflect this.

So the interesting question is likely to be (3).

Reply:
My analysis is similar to yours:  @Nullable on any class that makes a commitment to handle null, such as `ArrayList`, but not on classes that don't make a commitment (like most interfaces and abstract classes).

You are right that `ArrayList` should have `@Nullable` on its parameters.  Sorry for that mistake; I have corrected it.  (It looks like I overlooked it because it uses slightly different wording about null than other classes.)

> users can extend ArrayList and override this method to reject null inputs.
Subtypes of `ArrayList` must also permit null values.  If a subclass of `ArrayList` were to forbid null elements, it would violate a commitment made by `ArrayList`, would violate behavioral subtyping, and the Nullness Checker would yield an error when type-checking it.

Are there any other classes that are incorrectly annotated?